// @0xpolygonid/js-sdk/dist/esm/loaders/schema.js
import axios from 'axios';
import { create } from 'ipfs-http-client';
/**
 * loads schemas from ipfs and http sources
 *
 * @export
 * @beta
 * @class UniversalSchemaLoader
 * @implements implements ISchemaLoader interface
 */
export class UniversalSchemaLoader {
    constructor(ipfsUrl) {
        this.ipfsUrl = ipfsUrl;
    }
    /**
     * loads schema by URL
     *
     * @param {string} url - schema URL
     * @returns `Promise<Uint8Array>`
     */
    async load(url) {
        const l = getLoader(url, this.ipfsUrl);
        const schemaRes = await l.load(url);
        return schemaRes;
    }
}
/**
 * loads schemas from http source
 *
 * @export
 * @beta
 * @class HttpSchemaLoader
 * @implements implements ISchemaLoader interface
 */
export class HttpSchemaLoader {
    /**
     *
     *
     * @param {string} url - schema URL
     * @returns `Promise<Uint8Array>`
     */
    async load(url) {
        const resp = await axios.get(url, { responseType: 'arraybuffer' });
        return resp.data;
    }
}
/**
 * loads schemas from ipfs source
 *
 * @export
 * @beta
 * @class IpfsSchemaLoader
 * @implements implements ISchemaLoader interface
 */
export class IpfsSchemaLoader {
    /**
     * Creates an instance of IpfsSchemaLoader.
     * @param {string} url - host of the ipfs node
     */
    constructor(url) {
        this.url = url;
        this.client = create({ url: this.url });
    }
    /**
     * loads schema from ipfs by its identifier
     *
     * @param {string} url - schema IPFS identifier
     * @returns `Promise<Uint8Array>`
     */
    async load(url) {
        const uri = new URL(url);
        const schemaRes = this.client.cat(uri.host);
        let schemaBytes = Uint8Array.from([]);
        for await (const num of schemaRes) {
            schemaBytes = Uint8Array.from(num);
        }
        return schemaBytes;
    }
}
/**
 * returns loader for schemas based on the URL protocol
 *
 * @export
 * @param {string} url - schema url
 * @param {string} [ipfsConfigUrl] - ipfs node host url
 * @returns ISchemaLoader
 */
export function getLoader(url, ipfsConfigUrl) {
    const uri = new URL(url);
    switch (uri.protocol) {
        case 'http:':
        case 'https:':
            return new HttpSchemaLoader();
        case 'ipfs:':
            return new IpfsSchemaLoader(ipfsConfigUrl);
        default:
            throw new Error(`loader for ${uri.protocol} is not supported`);
    }
}
