// @0xpolygonid/js-sdk/dist/esm/storage/shared/identity-storage.js
/**
 * Implementation of the IIdentityStorage with KV data source
 *
 * @export
 * @beta
 * @class IdentityStorage
 * @implements implements IIdentityStorage interface
 */
export class IdentityStorage {
    /**
     * Creates an instance of IdentityStorage.
     * @param {IDataSource<Identity>} _identityDataSource - data source for identities
     * @param {IDataSource<Profile>} _profileDataSource - data source for profiles
     */
    constructor(_identityDataSource, _profileDataSource) {
        this._identityDataSource = _identityDataSource;
        this._profileDataSource = _profileDataSource;
    }
    async saveProfile(profile) {
        const profiles = this._profileDataSource.load();
        const identityProfiles = profiles.filter((p) => p.genesisIdentifier === profile.genesisIdentifier);
        const toSave = identityProfiles.length ? [...identityProfiles, profile] : [profile];
        for (let index = 0; index < toSave.length; index++) {
            const element = toSave[index];
            this._profileDataSource.save(element.id, element);
        }
    }
    async getProfileByVerifier(verifier) {
        const profile = this._profileDataSource.get(verifier, 'verifier');
        if (!profile) {
            throw new Error('profile not found');
        }
        return profile;
    }
    async getProfileById(profileId) {
        const profile = this._profileDataSource.get(profileId);
        if (!profile) {
            throw new Error('profile not found');
        }
        return profile;
    }
    async getProfilesByGenesisIdentifier(genesisIdentifier) {
        return this._profileDataSource.load().filter((p) => p.genesisIdentifier === genesisIdentifier);
    }
    async getAllIdentities() {
        return this._identityDataSource.load();
    }
    async saveIdentity(identity) {
        this._identityDataSource.save(identity.identifier, identity, 'identifier');
    }
    async getIdentity(identifier) {
        return this._identityDataSource.get(identifier, 'identifier');
    }
}
/**
 * storage key for identities
 *
 * @static
 */
IdentityStorage.identitiesStorageKey = 'identities';
/**
 * storage key for profiles
 *
 * @static
 */
IdentityStorage.profilesStorageKey = 'profiles';
