// @0xpolygonid/js-sdk/dist/cjs/storage/filters/jsonQuery.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.StandardJSONCredentialsQueryFilter = exports.FilterQuery = exports.createFilter = exports.resolvePath = exports.comparatorOptions = exports.SearchError = void 0;
/**
 * search errors
 *
 * @export
 * @enum {number}
 */
var SearchError;
(function (SearchError) {
    SearchError["NotDefinedQueryKey"] = "not defined query key";
    SearchError["NotDefinedComparator"] = "not defined comparator";
})(SearchError = exports.SearchError || (exports.SearchError = {}));
exports.comparatorOptions = {
    $noop: () => true,
    $eq: (a, b) => a === b,
    $in: (a, b) => b.includes(a),
    $nin: (a, b) => !b.includes(a),
    $gt: (a, b) => a > b,
    $lt: (a, b) => a < b,
    $ne: (a, b) => a !== b
};
/**
 * credential search path resolver
 *
 * @param {object} object - object to query
 * @param {string} path - given path
 * @param {*} [defaultValue=null]
 */
const resolvePath = (object, path, defaultValue = null) => path.split('.').reduce((o, p) => (o ? o[p] : defaultValue), object);
exports.resolvePath = resolvePath;
/**
 * filter creation factory
 *
 * @param {string} path - given query path
 * @param {*} operatorFunc - filter operation
 * @param {*} value - filter value
 * @param {*} [isReverseParams=false] - reversed params
 */
const createFilter = (path, operatorFunc, value, isReverseParams = false) => {
    if (!operatorFunc) {
        throw new Error(SearchError.NotDefinedComparator);
    }
    return (credential) => {
        const credentialPathValue = (0, exports.resolvePath)(credential, path);
        if (!credentialPathValue) {
            return false;
            // throw new Error(`Not found path - ${path} to credential`);
        }
        if (isReverseParams) {
            return operatorFunc(value, credentialPathValue);
        }
        return operatorFunc(credentialPathValue, value);
    };
};
exports.createFilter = createFilter;
/**
 * Filter for queries of credentialSubject with a json path e.g  birthday.date
 *
 *
 * @export
 * @beta
 * @class FilterQuery
 * @implements implements IFilterQuery interface
 */
class FilterQuery {
    /**
     * Creates an instance of FilterQuery.
     * @param {string} path
     * @param {FilterOperatorFunction} operatorFunc
     * @param {*} value
     * @param {boolean} [isReverseParams=false]
     */
    constructor(path, operatorFunc, value, isReverseParams = false) {
        this.path = path;
        this.operatorFunc = operatorFunc;
        this.value = value;
        this.isReverseParams = isReverseParams;
    }
    /** {@inheritdoc IFilterQuery} */
    execute(credential) {
        if (!this.operatorFunc) {
            throw new Error(SearchError.NotDefinedComparator);
        }
        const credentialPathValue = (0, exports.resolvePath)(credential, this.path);
        if (!credentialPathValue) {
            return false;
        }
        if (this.isReverseParams) {
            return this.operatorFunc(this.value, credentialPathValue);
        }
        return this.operatorFunc(credentialPathValue, this.value);
    }
}
exports.FilterQuery = FilterQuery;
/**
 * creates filters based on proof query
 * @param {ProofQuery} query - proof query
 * @returns {*}  {FilterQuery[]} - array of filters to apply
 */
const StandardJSONCredentialsQueryFilter = (query) => {
    return Object.keys(query).reduce((acc, queryKey) => {
        const queryValue = query[queryKey];
        switch (queryKey) {
            case 'claimId':
                return acc.concat(new FilterQuery('id', exports.comparatorOptions.$eq, queryValue));
            case 'allowedIssuers': {
                const [first] = queryValue || [];
                if (first && first === '*') {
                    return acc;
                }
                return acc.concat(new FilterQuery('issuer', exports.comparatorOptions.$in, queryValue));
            }
            case 'type':
                return acc.concat(new FilterQuery('type', exports.comparatorOptions.$in, queryValue, true));
            case 'context':
                return acc.concat(new FilterQuery('@context', exports.comparatorOptions.$in, queryValue, true));
            case 'credentialSubjectId':
                return acc.concat(new FilterQuery('credentialSubject.id', exports.comparatorOptions.$eq, queryValue));
            case 'schema':
                return acc.concat(new FilterQuery('credentialSchema.id', exports.comparatorOptions.$eq, queryValue));
            case 'skipClaimRevocationCheck':
                return acc;
            case 'credentialSubject': {
                const reqFilters = Object.keys(queryValue).reduce((acc, fieldKey) => {
                    const fieldParams = queryValue[fieldKey];
                    const res = Object.keys(fieldParams).map((comparator) => {
                        const value = fieldParams[comparator];
                        const path = `credentialSubject.${fieldKey}`;
                        return new FilterQuery(path, exports.comparatorOptions[comparator], value);
                    });
                    return acc.concat(res);
                }, []);
                return acc.concat(reqFilters);
            }
            default:
                throw new Error(SearchError.NotDefinedQueryKey);
        }
    }, []);
};
exports.StandardJSONCredentialsQueryFilter = StandardJSONCredentialsQueryFilter;
